### PR TITLE
dnsmasq: update the "Disable systemd-resolved" tasks

### PR DESCRIPTION
**Pull request by vitabaks**
Fixed:

ERROR: `src file does not exist`

```
TASK [consul : Disable systemd-resolved service] ****************************************************************************************************************************************************************
changed: [172.31.15.62]
changed: [3.223.140.234]
changed: [172.31.5.27]

TASK [consul : Check if resolv.conf is pointing to systemd-resolved] ********************************************************************************************************************************************
ok: [172.31.15.62]
ok: [3.223.140.234]
ok: [172.31.5.27]

TASK [consul : Remove resolv.conf association with systemd-resolved] ********************************************************************************************************************************************
fatal: [3.223.140.234]: FAILED! => {"changed": false, "gid": 0, "group": "root", "mode": "0777", "msg": "src file does not exist, use \"force=yes\" if you really want to create the link: /run/resolvconf/resolv.conf", "owner": "root", "path": "/etc/resolv.conf", "size": 39, "src": "/run/resolvconf/resolv.conf", "state": "link", "uid": 0}
fatal: [172.31.15.62]: FAILED! => {"changed": false, "gid": 0, "group": "root", "mode": "0777", "msg": "src file does not exist, use \"force=yes\" if you really want to create the link: /run/resolvconf/resolv.conf", "owner": "root", "path": "/etc/resolv.conf", "size": 39, "src": "/run/resolvconf/resolv.conf", "state": "link", "uid": 0}
fatal: [172.31.5.27]: FAILED! => {"changed": false, "gid": 0, "group": "root", "mode": "0777", "msg": "src file does not exist, use \"force=yes\" if you really want to create the link: /run/resolvconf/resolv.conf", "owner": "root", "path": "/etc/resolv.conf", "size": 39, "src": "/run/resolvconf/resolv.conf", "state": "link", "uid": 0}

NO MORE HOSTS LEFT **********************************************************************************************************************************************************************************************

PLAY RECAP ******************************************************************************************************************************************************************************************************
172.31.15.62               : ok=48   changed=19   unreachable=0    failed=1    skipped=68   rescued=0    ignored=0   
172.31.5.27                : ok=54   changed=21   unreachable=0    failed=1    skipped=69   rescued=0    ignored=1   
3.223.140.234              : ok=48   changed=19   unreachable=0    failed=1    skipped=68   rescued=0    ignored=0   
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0   

ubuntu@ip-172-31-5-27:~/postgresql_cluster$ ls -la /etc/resolv.conf
lrwxrwxrwx 1 root root 39 Dec  1 11:06 /etc/resolv.conf -> ../run/systemd/resolve/stub-resolv.conf
ubuntu@ip-172-31-5-27:~/postgresql_cluster$ ls -la /run/resolvconf/resolv.conf
ls: cannot access '/run/resolvconf/resolv.conf': No such file or directory
```

Test result:

```
ubuntu@source-pgnode02:~$ cat /etc/dnsmasq.d/10-consul
server=/consul/127.0.0.1#8600
server=8.8.8.8
server=8.8.4.4
ubuntu@source-pgnode02:~$ ls -la /etc/resolv.conf 
-rw-r--r-- 1 root root 21 Jan  4 19:22 /etc/resolv.conf
ubuntu@source-pgnode02:~$ 
ubuntu@source-pgnode02:~$ cat /etc/resolv.conf 
nameserver 127.0.0.1
ubuntu@source-pgnode02:~$ 
ubuntu@source-pgnode02:~$ dig +short master.patroni.service.consul SRV
1 1 6432 source-pgnode01.node.us-east-1d.consul.
```

### CODE REVIEW

**Comment by vitabaks** [comment]
Error for RedHat family version 9 

`FAILED! => {"changed": false, "msg": "Could not find the requested service systemd-resolved: host"}`

fixed: https://github.com/ansible-community/ansible-consul/pull/519/commits/f12246b50d4ba356117b7275a7b3a99ba8ec6f23

File: tasks/dnsmasq.yml
Comment (nre-ableton) [inline tasks/dnsmasq.yml]: Can you explain why CI failures are related to `unsafe_writes`? I'm not sure I see the connection.

File: tasks/dnsmasq.yml
Comment (nre-ableton) [inline tasks/dnsmasq.yml]: This (`stat.islnk`) will always be defined, assuming that the file exists. If the idea here is to check if the file exists before examining `stat.islnk` and `stat.lnk_source`, then I think this would be more readable:

```suggestion
        - resolv_dot_conf.stat.exists
```

**Review by nre-ableton — CHANGES_REQUESTED** [review]
Thanks for your PR! :+1: 

I don't understand what is going on with ff2cc4c3da4a51a5407eb5f0e1dc99ec39717792 and 102da4c46e8ad4cb47212bbaa0155d8e7d343329. The commit message of the former claims that this is necessary to fix compatibility with RedHat 9.x, but then the next commit reverts it without explanation.

If this fix is indeed not necessary, please remove the two commits from the branch. Otherwise, what is necessary to fix RH compatibility?

I also have a few other small comments on this PR, see below.

File: tasks/dnsmasq.yml
Comment (vitabaks) [inline tasks/dnsmasq.yml]:
```suggestion
```

File: tasks/dnsmasq.yml
Comment (vitabaks) [reply tasks/dnsmasq.yml]: this is necessary to avoid the error of the condition "resolv_dot_conf.stat.islnk" if there is no link information in the variable resolv_dot_conf

Comment (vitabaks) [reply tasks/dnsmasq.yml]: > If this fix is indeed not necessary, please remove the two commits from the branch. Otherwise, what is necessary to fix RH compatibility?

I don't remember the details, maybe I fixed my mistake here. \
I'm not sure I know how to remove commit from history except to create a PR again.

Comment (nre-ableton) [reply tasks/dnsmasq.yml]: > this is necessary to avoid the error of the condition "resolv_dot_conf.stat.islnk" if there is no link information in the variable resolv_dot_conf

Just out of curiosity, when does that happen? I've never seen `islnk` not be defined.

Comment (nre-ableton) [reply tasks/dnsmasq.yml]: > > If this fix is indeed not necessary, please remove the two commits from the branch. Otherwise, what is necessary to fix RH compatibility?
> 
> I don't remember the details, maybe I fixed my mistake here. I'm not sure I know how to remove commit from history except to create a PR again.

You would need to rebase the branch and rewrite history. I can try to do that for you, but GitHub is being weird and not letting me push, even though it claims that maintainers can edit the branch.

*Edit*: Ok, I figured out what was wrong. @vitabaks sorry for the force-push to your branch, but this is necessary when rewriting history. Please fetch my changes and pull them to your working copy so they don't get accidentally overwritten. :slightly_smiling_face:

**Review by nre-ableton — COMMENTED** [review]
Ok, this PR is starting to look much better. However, I'm a bit concerned about the lack of a molecule test. Is this something that can easily be tested in a Docker container? If not, can you test it locally on whatever platforms you have access to?

**Comment by vitabaks** [comment]
> Ok, this PR is starting to look much better. However, I'm a bit concerned about the lack of a molecule test. Is this something that can easily be tested in a Docker container? If not, can you test it locally on whatever platforms you have access to?

this has been tested as part of the project https://github.com/vitabaks/postgresql_cluster

I didn't do any molecule tests as part of this PR.

File: tasks/dnsmasq.yml
Comment (vitabaks) [reply tasks/dnsmasq.yml]: thanks @nre-ableton

File: tasks/dnsmasq.yml
Comment (nre-ableton) [inline tasks/dnsmasq.yml]:
```suggestion
        - name: Add a nameserver entry poining to localhost for dnsmasq
          lineinfile:
            path: /etc/resolv.conf
            regexp: "^nameserver 127.0.0.1"
            line: "nameserver 127.0.0.1"
            create: true
            owner: root
            group: root
            mode: "u=rw,g=r,o=r"
```

**Review by nre-ableton — COMMENTED** [review]
One final suggestion -- we can use the `create` property of `lineinfile`to avoid an unnecessary `file` task with `state: touch`.

File: tasks/dnsmasq.yml
Comment (nre-ableton) [inline tasks/dnsmasq.yml:55]: Note: we could also potentially change this to use `service_facts`, but let's consider that in a follow-up.

**Review by nre-ableton — DISMISSED** [review]
Ok, looks good! I'll go ahead and squash for you.

**Comment by nre-ableton** [comment]
@vitabaks Ok, this PR looks good. :+1: Could you please run a quick test with the latest changes before I merge?

File: tasks/dnsmasq.yml
Comment (vitabaks) [reply tasks/dnsmasq.yml:55]: I think it should be a separate PR

**Comment by vitabaks** [comment]
> Could you please run a quick test with the latest changes before I merge?

@nre-ableton Ok, here are the test results on my repository

image


details: https://github.com/vitabaks/postgresql_cluster/actions/runs/4303678548

**Comment by nre-ableton** [comment]
@vitabaks Great! Unfortunately some linter tasks are failing on our end. :disappointed: I'll investigate those this afternoon and try to come up with a fix.

**Review by nre-ableton — APPROVED** [review]

**Comment by nre-ableton** [comment]
@vitabaks thanks again for your contribution!

**Pull request merged**
